### ISSUE TITLE
LanguageClient failing to initialize language server on 1.62.1

### ISSUE THREAD

**Issue by bcanzanella** [microsoft/vscode#136988]
We're running into an issue that seems to have been introduced with 1.62.1. 

Upon initialization of our [LanguageClient](https://github.com/TeamCodeStream/codestream/blob/develop/vscode/src/agent/agentConnection.ts#L1069) in our VSCode extension, it's immediately triggering a closed notification on `LanguageClient.clientOptions.errorHandler.closed` ([here](https://github.com/TeamCodeStream/codestream/blob/develop/vscode/src/agent/agentConnection.ts#L284))

Insiders 1.63.0 also has the problem. It happens while debugging as well as when installing from the marketplace. 

If we downgrade to 1.61.2 or 1.60.2 there are no problems.

<img width="2008" alt="Screen Shot 2021-11-11 at 2 35 24 PM" src="https://user-images.githubusercontent.com/834948/141360042-3cfff1aa-d4b2-4a2b-8e13-d195ab6233c3.png">

- We checked the lsp-sample using 1.62.1 and using the same client + server libraries and there's no error.
- We changed our retry logic to retry indefinitely, and at some point it will complete. it seems like it's related to the `client.onReady()` promise never resolving [?]
- What else can we try, this is essentially affecting all our users.

Does this issue occur when all extensions are disabled?: N/A, this is inside a VSCode extension

Version: 1.62.1
Commit: f4af3cbf5a99787542e2a30fe1fd37cd644cc31f
Date: 2021-11-05T09:23:14.144Z
Electron: 13.5.2
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Darwin x64 20.6.0

Happens in both VSC 1.62.2 and VSCInsiders 1.63.0
Steps to Reproduce:

1. install CodeStream from the marketplace
2. close VSCode, then reopen it. 75% of the time you'll see a toast notification error

**Comment by ycrepeau** [comment]
Moved back from 1.62 to 1.61.2 and the problem is gone
![20211111_172322_Spectacle](https://user-images.githubusercontent.com/6821466/141377590-2d4de401-4b65-4479-8f46-78ecd4eaff16.jpeg)
.

**Comment by chrmarti** [comment]
Sorry about the breakage, this is from a fix to address an elevation of privilege vulnerability #136771 (CVE-2021-42322). Starting with 1.62.1 you need to pass the command line argument `--ms-enable-electron-run-as-node` to the VSCode binary in addition to `ELECTRON_RUN_AS_NODE=1` to use it as Node.js.

**Comment by bcanzanella** [comment]
@chrmarti could you clarify your solution for me? How / where are those additional arguments set?
We are a VSCode extension and have no ability to change how the user launched VSCode.

**Comment by jonathanm-tkf** [comment]
@chrmarti workaround for mac? thanks

**Comment by alexdima** [comment]
@bcanzanella It is not clear to me that this is a regression caused by our security fix, because our security fix changes Electron's behavior in 100% of cases, not just in 75% of cases.

Can you please clarify if this reproduces in `1.62.0` or not? Is `1.62.1` the first bad version?

**Comment by bcanzanella** [comment]
So the 75%/25% seems like an difference between running VSCode on a "read-only" volume, in my case from the Downloads folder rather than installed directly. in this case the error happens immediately 75% of the time (via "read-only" volume), and the other 25% seems to load, but later the LSP agent restarts multiple times and makes CodeStream unusable, but let's shelve that for the moment...

100% of the time the following is true:

1.62.0 WORKS   
1.62.1 BAD   
1.62.2 BAD

**Comment by alexdima** [comment]
@bcanzanella I have investigated and it looks like the forked process `dist/agent.js` exits with a `SIGSEGV`.

@deepak1556 Is there anything we can do to find out why this process crashes? AFAICT it is not loading any native node modules.

**Comment by atscott** [comment]
> @bcanzanella I have investigated and it looks like the forked process `dist/agent.js` exits with a `SIGSEGV`.
> 
> @deepak1556 Is there anything we can do to find out why this process crashes? AFAICT it is not loading any native node modules.

This is exactly the [issue](https://github.com/angular/angular/issues/44148) we are experiencing in the Angular Language Service when trying to run [`ngcc` as a forked process](https://github.com/angular/vscode-ng-language-service/blob/489d34d34e6fc6897db8b6080ddebbc0c1a56a96/server/src/ngcc.ts#L44-L48).  In `1.61.1`, `1.61.2`, and the insider version, the process fails with `SIGEGV` and a `null` code.

**Comment by deepak1556** [comment]
@bcanzanella when I repro the crash with codestream extension, the stacks indicate crash happens during compilation of the `dist/agent.js` file, the following is the translated callsite

```
"%cCHILD_PROCESS 66072: spawn {
%c  execArgv: [],
%c  cwd: '/Users/demohan/github/electron-quick-start',
%c  silent: true,
%c  stdio: [ 'pipe', 'pipe', 'pipe', 'ipc' ],
%c  env: { ELECTRON_RUN_AS_NODE: 1 },
%c  execPath: '/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer)',
%c  shell: false,
%c  args: [
%c    '/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer)',
%c    '--ms-enable-electron-run-as-node',
%c    '/Users/demohan/.vscode/extensions/codestream.codestream-12.1.0/dist/agent.js',
%c    '--node-ipc',
%c    '--clientProcessId=66072'
%c  ],
%c  detached: false,
%c  envPairs: [....],
%c  file: '/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer)',
%c  windowsHide: false,
%c  windowsVerbatimArguments: false
%c}
```

When I compare it with https://github.com/TeamCodeStream/codestream/blob/7c7d4d98f9d93d0ee70690d3c12093995823e60e/atom/lib/workspace/agent/connection.ts#L419-L423 the args don't match, can you share the code calling spawn ?

I was trying to play with the `--no-lazy` flag and the crash no longer repros, not sure why it makes a difference.

```sh
// crashes
ELECTRON_RUN_AS_NODE=1 /Applications/Visual\ Studio\ Code.app/Contents/Frameworks/Code\ Helper\ \(Renderer\).app/Contents/MacOS/Code\ Helper\ \(Renderer\) /Users/demohan/.vscode/extensions/codestream.codestream-12.1.0/dist/agent.js --node-ipc --ms-enable-electron-run-as-node

// does not crash
ELECTRON_RUN_AS_NODE=1 /Applications/Visual\ Studio\ Code.app/Contents/Frameworks/Code\ Helper\ \(Renderer\).app/Contents/MacOS/Code\ Helper\ \(Renderer\) /Users/demohan/.vscode/extensions/codestream.codestream-12.1.0/dist/agent.js --node-ipc --ms-enable-electron-run-as-node --no-lazy
```

**Comment by deepak1556** [comment]
@atscott can you share some minimal steps to repro the issue, thanks!

**Comment by bcanzanella** [comment]
@deepak1556 here is the code that spawns it: https://github.com/TeamCodeStream/codestream/blob/develop/vscode/src/agent/agentConnection.ts#L257

**Comment by bcanzanella** [comment]
also, @deepak1556 does the order of those arguments matter? in your debugging code `--ms-enable-electron-run-as-node` is before the script arg (agent.js)... 

```
args: [
%c    '/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer)',
%c    '--ms-enable-electron-run-as-node',
%c    '/Users/demohan/.vscode/extensions/codestream.codestream-12.1.0/dist/agent.js',
%c    '--node-ipc',
%c    '--clientProcessId=66072'
%c  ],
```

...but in your manual testing the `--ms-enable-electron-run-as-node` arg is after agent.js

**Comment by deepak1556** [comment]
Order doesn't matter `--ms-enable-electron-run-as-node` argument is parsed and removed early on in the process startup.

**Comment by atscott** [comment]
Hi @deepak1556, sorry if these aren't minimal enough directions! Let me know what I can do to provide something better:

1. Clone our [extension repo](https://github.com/angular/vscode-ng-language-service)
2. Open the repository with vscode `1.62.1` or above
3. Install dependencies with `yarn && cd integration/project/ && yarn`
4. Run the build script with `./scripts/build.sh`
5. For debugging, put a breakpoint on the [line that calls `fork`](https://github.com/angular/vscode-ng-language-service/blob/master/server/src/ngcc.ts#L44) and one on the [`close listener`](https://github.com/angular/vscode-ng-language-service/blob/master/server/src/ngcc.ts#L69)
6. Launch the extension development host (`Dev Client + Server` run configuration)
7. In the extension host, open the [integration project](https://github.com/angular/vscode-ng-language-service/tree/master/integration/project) and then open the `app.component.html` template file
 -- debugger should stop on the `fork` and if you continue, you will see that the debugger stops in the `childProcess.on('close')` with a `null` code.
You can have the `fork` retry by running the `Angular: Restart Angular Language server` command in the extension host.

**Comment by atscott** [comment]
I also attempted to add the `--ms-enable-electron-run-as-node` flag to the `child_process` `fork` but that didn't work. Also, FWIW, I _was_ able to add a `spawn('pwd');` that succeeded with `code 0` at the same location.

**Comment by deepak1556** [comment]
End-to-end debugging of the issue narrowed down the problem to a UAF scenario in the argument parser. I am getting a new electron build ready with the fix today. I will update here once the fix is rolled into our insiders for some testing. Sorry about the breakage!

**Comment by deepak1556** [comment]
For anyone available for early testing, here are some test builds.

[macOS x64](https://az764295.vo.msecnd.net/insider/5be35751b91c4807756ada25f0dff609c387c014/VSCode-darwin.zip)
[windows x64](https://az764295.vo.msecnd.net/insider/5be35751b91c4807756ada25f0dff609c387c014/VSCode-win32-x64-1.63.0-insider.zip)
[linux x64](https://az764295.vo.msecnd.net/insider/5be35751b91c4807756ada25f0dff609c387c014/code-insider-x64-1636980599.tar.gz)

**Comment by bcanzanella** [comment]
thanks @deepak1556 we have confirmation from a few CodeStream developers that it is fixed in the macOS version

**Comment by atscott** [comment]
@deepak1556 I can also confirm that this resolves the issue with forking a process for `ngcc` in the Angular language service extension. Thank you!

**Comment by deepak1556** [comment]
Today's insider Version: 1.63.0-insider (Universal) 6a25ae3a24929f35aabf8a051531c0716032e0a0 , carries the fix. I will close this optimistically based on the testing. The fix will also be made available in recovery release `1.62.3` later this week.
